ipinfo/cmd_download.go

```diff
@@ -0,0 +1,160 @@
+)
+
+const dbDownloadURL = "https://ipinfo.io/data/free/"
+
+var completionsDownload = &complete.Command{
+	Flags: map[string]complete.Predictor{
+		"--asn":         predict.Nothing,
+		"--country":     predict.Nothing,
+		"--country-asn": predict.Nothing,
+		"-f":            predict.Nothing,
+		"--format":      predict.Nothing,
+		"-t":            predict.Nothing,
+		"--token":       predict.Nothing,
+		"-h":            predict.Nothing,
+		"--help":        predict.Nothing,
+	},
+}
+
+func printHelpDownload() {
+	fmt.Printf(
+		`Usage: %s download [<opts>]
+
+Description:
+    Download the free ipinfo databases.
+
+Examples:
+    # Download country database in csv format.
+    $ %[1]s download --country --csv
+
+Options:
+  General:
+    --token <tok>, -t <tok>
+      use <tok> as API token.
+    --asn
+      download the free ipinfo asn database. 
+    --country
+      download the free ipinfo country database. 
+    --country-asn
+      download the free ipinfo country asn database. 
+    --format <fmt>, -f <fmt>
+      use the <fmt> output format of the database file.
+        mmdb (default.)
+        json
+        csv
+    --help, -h
+      show help.
+`, progBase)
+}
+
+func cmdDownload() error {
+	var fTok string
+	var fFmt string
+	var fAsn bool
+	var fCountry bool
+	var fCountryAsn bool
+	var fHelp bool
+
+	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
+	pflag.BoolVar(&fAsn, "asn", false, "free asn database.")
+	pflag.BoolVar(&fCountry, "country", false, "free country database.")
+	pflag.BoolVar(&fCountryAsn, "country-asn", false, "free country asn database.")
+	pflag.StringVarP(&fFmt, "format", "f", "mmdb", "the output format to use.")
+	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.Parse()
+
+	if pflag.NFlag() == 0 {
+		printHelpDownload()
+		return nil
+	}
+
+	if fHelp {
+		printHelpDownload()
+		return nil
+	}
+
+	var token string
+	if fTok == "" {
+		token = gConfig.Token
+	}
+
+	// require token for download.
+	if token == "" {
+		return errors.New("download require a token; login via `ipinfo login` or pass the `--token` argument")
+	}
+
+	// check download format
+	var format string
+	switch fFmt {
+	case "mmdb":
+		format = "mmdb"
+	case "csv":
+		format = "csv.gz"
+	case "json":
+		format = "json.gz"
+	default:
+		return errors.New("unknown download format")
+	}
+
+	if fAsn {
+		err := downloadDb("asn", format, token)
+		if err != nil {
+			return err
+		}
+	}
+	if fCountry {
+		err := downloadDb("country", format, token)
+		if err != nil {
+			return err
+		}
+	}
+	if fCountryAsn {
+		err := downloadDb("country_asn", format, token)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func downloadDb(name, format, token string) error {
+	url := fmt.Sprintf("%s%s.%s?token=%s", dbDownloadURL, name, format, token)
+
+	// make API req to download the file.
+	res, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// create file.
```

Comment:
Our CLI is super flexible and needs to be remain so. We should be detecting if `stdout` is connected to a pipe/`>` (I dont know how to check - we do it for `stdin` though), and if so outputting to `stdout` so the user gets what they wanted. Only if that isn't happening then we can create a file. We should also optionally provide a positional argument as the last argument which specifies the output pathname.

Reply:
The issue with allowing the user to save the output or give the optional argument for the name is that 2 out of 3 files are in `.gz` format and not directly `json` and `csv` so that's why I didn't allow that option so the user would know and extract the files manually. If we allow saving with a custom name the user might mislead and save the files like `ipinfo download country -f csv > country.csv` and then try to open that file with some csv editor and that'll not work.

Reply:
I don't think there's a way to check `stdout` as even outputting on terminal is `stdout` if we are going with `3rd argument (optional)` then we can go with a hacky approach and check if the 3rd argument is `>` then we can take the 4th argument as file name but again that's a hacky approach and again I said earlier the files are in `gz` format so allowing the redirect will not make sense as we'll not be supporting `|` anyway.

Reply:
See https://unix.stackexchange.com/questions/515778/how-does-a-program-know-if-stdout-is-connected-to-a-terminal-or-a-pipe

Reply:
In terms of compression, we can use golang pipes (see io package, I think it has some fancy utilities there) to pipe the response body into a gzipper to unzip and stream that into a file.

We can then intro a flag called `-c, --compress` which would skip this unzipping streaming step and leave it downloaded as zipped. The user knows it's compressed because they passed this file in.

So all remains intuitive but at the same time flexible - we 100% should be able to put this anywhere we want on the filesystem, it shouldn't be forced to be right in the same directory. It's against the ethos of this flexible CLI. We gotta find solutions to make it flexible all the time.